src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java

```diff
@@ -0,0 +1,156 @@
+package de.tum.in.www1.artemis.service;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import javax.annotation.PostConstruct;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+
+import de.tum.in.www1.artemis.domain.Organization;
+import de.tum.in.www1.artemis.domain.User;
+import de.tum.in.www1.artemis.repository.OrganizationRepository;
+import de.tum.in.www1.artemis.repository.UserRepository;
+import de.tum.in.www1.artemis.security.SecurityUtils;
+import de.tum.in.www1.artemis.web.rest.errors.EntityNotFoundException;
+
+/**
+ * Service implementation for managing Organization entities
+ */
+@Service
+public class OrganizationService {
+
+    private final Logger log = LoggerFactory.getLogger(OrganizationService.class);
+
+    private final OrganizationRepository organizationRepository;
+
+    private final UserRepository userRepository;
+
+    public OrganizationService(OrganizationRepository organizationRepository, UserRepository userRepository) {
+        this.organizationRepository = organizationRepository;
+        this.userRepository = userRepository;
+    }
+
+    /**
+     * Called at application startup, this method tries to assign organizations to
+     * users which do not belong to any. For each user without organization, the email
+     * patterns of the organization are used to find any matching, resulting in the assignment
+     * of the user to the matcher organization.
+     */
+    @PostConstruct
+    public void init() {
```

Comment:
This should only be invoked once.
Also this does not really support multi node environments

Reply:
I suggest to let the system admin invoke this functionality with a button.

Reply:
I suggest Artemis shows a warning and the user needs to confirm this operation

Reply:
* PostConstruct part (indexing at application startup) has been removed
* Indexing implementation has been adapted to the suggestion (using the new **findAllMatchingEmailPattern** method)
* Indexing now uses 'strict' indexing (removes users which do not match the organization
* I assume the warning for the user is not needed as he/she cannot change his/her own email. Only admins can. (see screenshot)
![Screenshot from 2021-03-16 09-06-32](https://user-images.githubusercontent.com/45761921/111275974-16df7a80-8637-11eb-9b46-42d97d1ec064.png)